src/Npgsql/NpgsqlWriteBuffer.cs

```diff
@@ -155,13 +155,16 @@ public async Task Flush(bool async, CancellationToken cancellationToken = defaul
                     Underlying.Flush();
                 }  
             }
+            // User requested the cancellation
+            catch (OperationCanceledException) when (cancellationToken.IsCancellationRequested)
```

Comment:
It seems like we avoid breaking the connector here, only to break it up the stack (in ExecuteReader). Any reason to not let the connector just be broken here?

Reply:
It's not up the stack - don't forget, that the send task (where we use the `WriteBuffer`) is only partially sync. After that, we read and write in parallel.

The main idea here was to cancel the write side the moment we've send the cancellation request. And after that, we only have to deal with the read side. Then, we either read a response from PG, or fail with the cancellation timeout.

Reply:
I see... You want to cancel the write, but if it's in the async part, you want to leave the breaking of the connector to be done by the read side, is that right?

If so, that makes sense to me - but there's one problem... Unless I'm mistaken, if we actually break an ongoing write (sync or async), the connector becomes unusable (we don't know what got written and what hasn't, so we lose protocol sync). But as the code currently is, we just send a PG cancellation - which may actually succeed - but the next query would fail because of partial garbage, no? If that's correct, then as soon as *any* write is cancelled, we should just break, no?

As this stuff seems complicated, it may be good to leave all the async write issues for handling as part of #1323.

Reply:
It should be fine, as we're clearing the write buffer here:
https://github.com/npgsql/npgsql/blob/a570da94db0e7be3625a024bfbaf3137ad87ad8f/src/Npgsql/NpgsqlConnector.cs#L1705-L1717

But I do agree, that this stuff is very complicated (and it's more of an optimization), and can be done as part of #1323.
For now, we just go with the simplest solution - breaking the connection.